topics/single-cell/tutorials/pseudobulk-analysis/tutorial.md

```diff
@@ -0,0 +1,503 @@
+>    - {% icon param-file %} *"Tabular file"*: `genes_metadata` (output of **Decoupler pseudo-bulk** {% icon tool %})
+>    - In *"Select Columns"*:
+>        - {% icon param-repeat %} *"Insert Select Columns"*
+>            - *"Header name"*: `start`
+>        - {% icon param-repeat %} *"Insert Select Columns"*
+>            - *"Header name"*: `end`
+>        - {% icon param-repeat %} *"Insert Select Columns"*
+>            - *"Header name"*: `width`
+>
+{: .hands_on}
+
+> <hands-on-title> Replace Text: Creating Factor File </hands-on-title>
+>
+> 1. {% tool [Replace Text](toolshed.g2.bx.psu.edu/repos/bgruening/text_processing/tp_replace_in_column/9.3+galaxy1) %} with the following parameters:
+>    - {% icon param-file %} *"File to process"*: `outfile` (output of **Replace Text: Sample Metadata Step** {% icon tool %})
+>    - In *"Replacement"*:
+>        - {% icon param-repeat %} *"Insert Replacement"*
+>            - *"in column"*: `c2`
+>            - *"Find pattern"*: `^([0-9])(.+)`
+>            - *"Replace with"*: `GG_\\1\\2`
+>
+{: .hands_on}
+
+## Generating the Contrast File 
+
+This file will be used as the contrast input file in the edgeR tool. 
+
+> <hands-on-title> Creating a Contrast File for edgeR </hands-on-title>
+>
+> 1. {% tool [Text Reformatting](toolshed.g2.bx.psu.edu/repos/bgruening/text_processing/tp_awk_tool/9.3+galaxy1) %} with the following parameters:
+>    - {% icon param-file %} *"File to process"*: `outfile` (output generated from the **Replace Text** {% icon tool %} step).
+>    - *"AWK Program"*:
+>      ```awk
+>      BEGIN { print header } 
+>      NR > 1 { if (!seen[$2]++) words[++count]=$2 } 
+>      END { for (i=1; i<=count; i++) 
+>               for (j=i+1; j<=count; j++) 
+>                  print words[i]-words[j] }
+>      ```
+> 
+>    > <comment-title> Explanation of the AWK Program </comment-title>
+>    >
+>    > This AWK script performs the following:
+>    > - Initializes the `header` variable for the output file.
+>    > - Processes the input file (`NR > 1` skips the header line).
+>    > - Tracks unique elements in column 2 (`seen[$2]++`) to avoid duplicates.
+>    > - Creates pairs of unique elements and calculates their difference, outputting the contrast for downstream edgeR analysis.
+>    {: .comment}
+>
+{: .hands_on}
+
+> <question-title> Understanding the Contrast File </question-title>
+>
+> 1. How does your contrast file look, and what contrast is generated by the tool?
+>
+> <solution-title> Explanation of the Contrast File </solution-title>
+> 
+> 1. The contrast file is a simple tab-delimited text file. It contains:
+>    - A **header** in the first row that labels the column.
+>    - The contrast, such as `normal-COVID_19`, written in the second row.
+>
+> When working with your own data, the contrast file will look different, as it will reflect the specific contrasts in your dataset but the **header** should be the same
```

Comment:
This box is not rendering well, the solution box needs extra indentation 

```suggestion
> > <solution-title> Explanation of the Contrast File </solution-title>
> >
> > 1. The contrast file is a simple tab-delimited text file. It contains:
> >    - A **header** in the first row that labels the column.
> >    - The contrast, such as `normal-COVID_19`, written in the second row.
> >
> > When working with your own data, the contrast file will look different, as it will reflect the specific contrasts in your dataset but the **header** should be the same
```